sdk/src/trace/probability_sampler.cc

```diff
@@ -0,0 +1,127 @@
+
+#include "opentelemetry/sdk/trace/probability_sampler.h"
+
+#include <cmath>
+#include <cstdint>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+ProbabilitySampler::ProbabilitySampler(
+  double probability, bool defer_parent, SamplingBehavior sampling_behavior 
+)
+: threshold_(CalculateThreshold(probability)), probability_(probability),
+  defer_parent_(defer_parent), sampling_behavior_(sampling_behavior) {}
+
+SamplingResult ProbabilitySampler::ShouldSample(
+  const SpanContext *parent_context,
+  trace_api::TraceId trace_id,
+  nostd::string_view name,
+  trace_api::SpanKind span_kind,
+  const trace_api::KeyValueIterable &attributes) noexcept
+{
+  // Check if current Span should be sampled with current Sampling Behavior
+  if (sampling_behavior_ == SamplingBehavior::DETACHED_SPANS_ONLY
+      && parent_context != nullptr && !parent_context->is_remote) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  if (sampling_behavior_ == SamplingBehavior::ROOT_SPANS_ONLY
+      && parent_context != nullptr) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  // Check if sampling decision should be defered to parent span, if applicable 
+	if (parent_context == nullptr || !defer_parent_)
+  {
+    if (threshold_ == 0) return { Decision::NOT_RECORD, nullptr };
+
+    // Calculcate probability based decision based on trace_id and threshold
+	  if (CalculateThresholdFromBuffer(trace_id) <= threshold_)
+    {
+      return { Decision::RECORD_AND_SAMPLE, nullptr };
+    }
+	}
+  else
+  {
+    // Return sampling decision based on parent span_context config
+    if (parent_context->sampled)
+    {
+      return { Decision::RECORD_AND_SAMPLE, nullptr };
+    }
+	}
+
+  return { Decision::NOT_RECORD, nullptr };
+}
+
+/**
+ * @return Description MUST be ProbabilitySampler{0.000100}
+ */
+std::string ProbabilitySampler::GetDescription() const noexcept
+{
+	char buffer[30];
+	sprintf(buffer, "ProbabilitySampler{%.6f}", GetProbability());
+  return std::string(buffer);
+}
+
+uint64_t ProbabilitySampler::CalculateThreshold(double probability) const noexcept
+{
+  if (probability <= 0.0) return 0;
+  if (probability >= 1.0) return UINT64_MAX;
+  // We can't directly return probability * UINT64_MAX.
+  //
+  // UINT64_MAX is (2^64)-1, but as a double rounds up to 2^64.
+  // For probabilities >= 1-(2^-54), the product wraps to zero!
+  // Instead, calculate the high and low 32 bits separately.
+  const double product = UINT32_MAX * probability;
+  double hi_bits, lo_bits = ldexp(modf(product, &hi_bits), 32) + product;
+  return (static_cast<uint64_t>(hi_bits) << 32) +
+         static_cast<uint64_t>(lo_bits);
+}
+
+uint64_t ProbabilitySampler::CalculateThresholdFromBuffer(const trace_api::TraceId& trace_id) const noexcept
+{
+  // We only use the first 8 bytes of TraceId.
+  static_assert(trace_api::TraceId::kSize >= 8, "TraceID must be at least 8 bytes long.");
```

Comment:
static_assert will only throw an error at compile time if the case is not met. It shouldn't be a very easy error to trigger, given that TraceId::kSize is a static constant that shouldn't be touched by end users. However, if someone changes the size of the Trace in the future to something under 8 bytes, using your suggestion, this function will start return throwing an value that will still allow some of the test cases to pass and could be tricky to isolate the problem. Even worse, if the test cases for this class are modified in the future, there is a chance that even with returning an invalid value (i.e. -1) all test cases could pass and this could ship to production

Reply:
If static_assert is a compile time check, then this is totally fine. My suggestion is that we want to avoid crashing the service, even if the probability sampler needs to fail for it to happen (in system design lingo this is called "fail-open").